### PR TITLE
🐛Api-server wrong start/stop/submit timestamps

### PR DESCRIPTION
**Pull request by sanderegg**
## What do these changes do?
fixes https://github.com/ITISFoundation/osparc-simcore/issues/4231

| timestamp | BEFORE | AFTER |
|--------|--------|--------|
| submitted | request local now time | last modification of pipeline |
| started_at | timestamp of inspect request when pipeline state was started (very unreliable) | start timestamp when director-v2 submits the job to the comp. backend |
| stopped_at | timestamp of inspect request when pipeline state was stopped (very unreliable and tends to be the latest time it was called) | stop timestamp when director-v2 receives pipeline completed (abort, fail or success) |

image

Bonuses:
- improves testing by using pydantic models to detect missing variables
- some SQLAlchemy 2.0 updates

### CODE REVIEW

File: services/api-server/src/simcore_service_api_server/models/schemas/jobs.py [hunk 1]
```diff
@@ -231,7 +218,7 @@ class JobStatus(BaseModel):
     #  SEE https://english.stackexchange.com/questions/12958/status-vs-state
 
     job_id: UUID
-    state: TaskStates
+    state: RunningState
```
Comment (pcrespov) [inline services/api-server/src/simcore_service_api_server/models/schemas/jobs.py:221]: THOUGHT:  Reusing entities (i.e. models, dataclasses, etc ) from a common library (like models_library) has definitively many advantages but at least one disadvantage: a change on e.g. RunningState affects directly the public API. I wonder if there is a good strategy to control any unintended change in the public APIs.

File: services/api-server/src/simcore_service_api_server/utils/solver_job_models_converters.py [hunk 2]
```diff
@@ -252,19 +246,9 @@ def create_jobstatus_from_task(task: ComputationTaskGet) -> JobStatus:
         job_id=task.id,
         state=task.state,
         progress=PercentageInt((task.pipeline_details.progress or 0) * 100.0),
-        submitted_at=datetime.utcnow(),
+        submitted_at=task.submitted or arrow.utcnow().datetime,
```
Comment (pcrespov) [inline services/api-server/src/simcore_service_api_server/utils/solver_job_models_converters.py:249]: Q: why would this happen? shouldn't ``task.submitted`` be not `None` when the task is submitted? 
Moreover, can it happen that either `started_at` or `stopped_at` < `submitted_at` if `submitted_at=arrow.utcnow().datetime` ?

File: services/web/server/src/simcore_service_webserver/projects/projects_db.py [hunk 3]
```diff
@@ -276,21 +276,21 @@ async def list_projects(
...
             query = (
-                sa.select([projects, projects_to_products.c.product_name])
+                sa.select(projects, projects_to_products.c.product_name)
                 .select_from(projects.join(projects_to_products, isouter=True))
                 .where(
                     (
                         (projects.c.type == filter_by_project_type.value)
                         if filter_by_project_type
-                        else (projects.c.type != None)
+                        else (projects.c.type.is_not(None))
```
Comment (pcrespov) [inline services/web/server/src/simcore_service_webserver/projects/projects_db.py:285]: is this 2.0 style? noted :-)

File: services/api-server/src/simcore_service_api_server/models/schemas/jobs.py [hunk 4]
```diff
@@ -252,14 +239,10 @@ class Config(BaseConfig):
         schema_extra = {
             "example": {
                 "job_id": "145beae4-a3a8-4fde-adbb-4e8257c2c083",
-                "state": TaskStates.STARTED,
+                "state": RunningState.STARTED,
                 "progress": 3,
                 "submitted_at": "2021-04-01 07:15:54.631007",
```
Comment (pcrespov) [inline services/api-server/src/simcore_service_api_server/models/schemas/jobs.py:244]: Can you please add the new definitions of submitted_at, started_at and stopped_at  (as written in the PR description) as descriptions to these fields int he `Job` model?

**Review by pcrespov — APPROVED** [review]
Excellent! Thx for the addition!

**Review by GitHK — APPROVED** [review]
👍

File: services/api-server/src/simcore_service_api_server/utils/solver_job_models_converters.py (hunk 2 above)
Comment (sanderegg) [reply services/api-server/src/simcore_service_api_server/utils/solver_job_models_converters.py:249]: submitted can be none if there are no computational services in the pipeline.

File: services/director-v2/src/simcore_service_director_v2/utils/dags.py [hunk 5]
```diff
@@ -179,6 +181,56 @@ async def create_minimal_computational_graph_based_on_selection(
...
 
+def compute_pipeline_started_timestamp(
+    pipeline_dag: nx.DiGraph, comp_tasks: list[CompTaskAtDB]
+) -> datetime.datetime | None:
+    if not pipeline_dag.nodes:
+        return None
+    node_id_to_comp_task: dict[NodeIDStr, CompTaskAtDB] = {
+        NodeIDStr(f"{task.node_id}"): task for task in comp_tasks
+    }
+    TOMORROW = arrow.utcnow().shift(days=1).datetime
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply services/director-v2/src/simcore_service_director_v2/utils/dags.py:192]: no that is more for checking whether there is actually a date defined or not.

File: services/api-server/src/simcore_service_api_server/models/schemas/jobs.py (hunk 1 above)
Comment (sanderegg) [reply services/api-server/src/simcore_service_api_server/models/schemas/jobs.py:221]: ok I see your point with the API. nevertheless, it was a painful red in my IDE. also there was no test checking that these were the same so I thought it was unnecessary.

File: packages/models-library/requirements/_base.in [hunk 6]
```diff
@@ -3,5 +3,6 @@
 #
 --constraint ../../../requirements/constraints.txt
 
+arrow
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply packages/models-library/requirements/_base.in:6]: will do it now. but I think arrow is already everywhere. the issue was that it was just not in models-lib

**Pull request merged**